### ISSUE TITLE
libopenblas.so misses dependency declaration on libflang.so when compiling with flang

### ISSUE THREAD

**Issue by leezu** [xianyi/OpenBLAS#2650]
Steps to reproduce

``` bash
cd /usr/local
wget https://github.com/flang-compiler/flang/releases/download/flang_20190329/flang-20190329-x86-70.tgz
tar xf flang-20190329-x86-70.tgz
make DYNAMIC_ARCH=1 USE_OPENMP=1 FC=flang
```

We can check the undefined symbols

```
nm -u libopenblas.so
                 U GOMP_parallel@@GOMP_4.0
                 w _ITM_deregisterTMCloneTable
                 w _ITM_registerTMCloneTable
                 U __assert_fail@@GLIBC_2.2.5
                 w __cxa_finalize@@GLIBC_2.2.5
                 U __errno_location@@GLIBC_2.2.5
                 U __fd_atan2_1
                 U __fd_cos_1
                 U __fd_cos_2
                 U __fd_exp_1
                 U __fd_log10_1
                 U __fd_log_1
                 U __fd_log_2
                 U __fd_pow_1
                 U __fd_powi_1
                 U __fd_sin_1
                 U __fd_sincos_1
                 U __fprintf_chk@@GLIBC_2.3.4
                 U __fs_atan2_1
                 U __fs_cos_1
                 U __fs_cos_4
                 U __fs_exp_1
                 U __fs_log10_1
                 U __fs_log_1
                 U __fs_log_4
                 U __fs_pow_1
                 U __fs_powi_1
                 U __fs_powi_4
                 U __fs_sin_1
                 U __fs_sin_4
                 U __fs_sincos_1
                 U __get_size_of
                 w __gmon_start__
                 U __kmpc_end_master
                 U __kmpc_fork_call
                 U __kmpc_global_thread_num
                 U __kmpc_master
                 U __kmpc_omp_task
                 U __kmpc_omp_task_alloc
                 U __memset_chk@@GLIBC_2.3.4
                 U __mth_i_cdabs
                 U __mth_i_cddiv
                 U __mth_i_cdiv
                 U __mth_i_cdsin
                 U __mth_i_cdsqrt
                 U __mth_i_csin
                 U __mth_i_csqrt
                 U __mth_i_idnint
                 U __mth_i_nint
                 U __printf_chk@@GLIBC_2.3.4
                 U __register_atfork@@GLIBC_2.3.2
                 U __sched_cpualloc@@GLIBC_2.7
                 U __sched_cpucount@@GLIBC_2.6
                 U __sched_cpufree@@GLIBC_2.7
                 U __snprintf_chk@@GLIBC_2.3.4
                 U __sprintf_chk@@GLIBC_2.3.4
                 U __stack_chk_fail@@GLIBC_2.4
                 U __strcat_chk@@GLIBC_2.3.4
                 U __tls_get_addr@@GLIBC_2.3
                 U __vfprintf_chk@@GLIBC_2.3.4
                 U ceilf@@GLIBC_2.2.5
                 U exit@@GLIBC_2.2.5
                 U f90_set_intrin_type_i8
                 U f90_str_copy_klen
                 U f90_str_cpy1
                 U f90_strcmp_klen
                 U fort_maxloc_i8
                 U fputs@@GLIBC_2.2.5
                 U free@@GLIBC_2.2.5
                 U fwrite@@GLIBC_2.2.5
                 U getenv@@GLIBC_2.2.5
                 U malloc@@GLIBC_2.2.5
                 U memcpy@@GLIBC_2.14
                 U memset@@GLIBC_2.2.5
                 U mmap@@GLIBC_2.2.5
                 U munmap@@GLIBC_2.2.5
                 U omp_get_max_threads@@OMP_1.0
                 U omp_get_num_threads@@OMP_1.0
                 U omp_get_num_threads_@@OMP_1.0
                 U omp_get_thread_num@@OMP_1.0
                 U omp_get_thread_num_@@OMP_1.0
                 U omp_in_parallel@@OMP_1.0
                 U omp_set_num_threads@@OMP_1.0
                 U perror@@GLIBC_2.2.5
                 U pthread_mutex_lock@@GLIBC_2.2.5
                 U pthread_mutex_unlock@@GLIBC_2.2.5
                 U puts@@GLIBC_2.2.5
                 U sched_getaffinity@@GLIBC_2.3.4
                 U sched_yield@@GLIBC_2.2.5
                 U sqrt@@GLIBC_2.2.5
                 U sqrtf@@GLIBC_2.2.5
                 U stderr@@GLIBC_2.2.5
                 U strncasecmp@@GLIBC_2.2.5
                 U strncpy@@GLIBC_2.2.5
                 U strtol@@GLIBC_2.2.5
                 U syscall@@GLIBC_2.2.5
                 U sysconf@@GLIBC_2.2.5
```

and the declared dependencies

```
readelf -d libopenblas.so | grep NEEDED
 0x0000000000000001 (NEEDED)             Shared library: [libm.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libpthread.so.0]
 0x0000000000000001 (NEEDED)             Shared library: [libgomp.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libc.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [ld-linux-x86-64.so.2]
```

`libflang.so` which provides some of the undefined symbols is erroneously not declared as dependency.

**Comment by leezu** [comment]
`libopenblas.a` also misses some symbols

```
undefined reference to `__fs_log_1'
undefined reference to `__fs_pow_1'
undefined reference to `__fs_log10_1'
undefined reference to `__mth_i_nint'
undefined reference to `__fd_log_1'
undefined reference to `__fs_powi_1'
undefined reference to `__fd_powi_1'
undefined reference to `f90_str_copy_klen'
undefined reference to `__fd_log10_1'
undefined reference to `__fd_pow_1'
undefined reference to `f90_strcmp_klen'
```

**Comment by martin-frbg** [comment]
The .so probably needs the final link done with FC instead of CC in exports/Makefile (like it is done with ifort, probably for similar reasons). The missing symbols in the .a are most likely from the various flang support libraries that come in addition to libflang itself (these probably did not exist in earlier versions of LLVM yet)

**Comment by leezu** [comment]
Thanks @martin-frbg. Which lines in exports/Makefile are you referring to? I'm not able to find ifort in that file.

Another concern is that the `DEPEND` clause used in a couple of Fortran file doesn't seem to be implemented in flang yet..: `F90-W-0547-OpenMP feature, DEPEND, not yet implemented in this version of the compiler. (dsytrd_sb2st.F: 486)`. And in fact some of the tests are failing..

**Comment by martin-frbg** [comment]
https://github.com/xianyi/OpenBLAS/blob/430e8b45fe9b2ad6aae35912fbccc74dc74b2104/exports/Makefile#L153-L157
and perhaps the changes from my very recent PR #2647 apply to all current versions of flang and not just its AMD fork - LAPACK certainly needs to be built with -frecursive, and I had to drop the
flang optimization level to -O1 to make all tests in test/ctest pass.
(And if flang does not support OMP DEPEND, it probably should not advertise OpenMP 4.0 capability - these have been "#if defined(_OPENMP) &&  _OPENMP >= 201307L" for quite some time in OpenBLAS. Though I suspect flang will just do the same i.e. ignore the "TASK DEPEND" section, even if the ifdef condition is not met)

**Comment by leezu** [comment]
Thank you @martin-frbg. I opened a PR adding a flang compilation step to Github Actions that reproduces the test failures (first commit; https://github.com/xianyi/OpenBLAS/pull/2651/checks?check_run_id=752820819) and tries your suggestion of enforcing O1 (second commit; https://github.com/leezu/OpenBLAS/runs/752820731).

One side-question: Is it possible to configure the build to fail once `FATAL ERROR - COMPUTED RESULT IS LESS THAN HALF ACCURATE` occurs for one of the tests?

**Comment by martin-frbg** [comment]
IIRC for the level 2 and 3 BLAS tests you can change a flag in the header of the respective input file (sblat2.dat etc) to stop on error. The alterative would be to grep the output files (the .SUMM), I think the cmake build already does this to be able to flag the error in the build log. These tests are copied from Reference-LAPACK and it is usually more interesting to know all the failing tests rather than just that at least one of them failed.

**Comment by martin-frbg** [comment]
Makefile build is definitely fixed by changing the "FLANG_VENDOR==AOCC" conditional in #2647 to a generalized "F_COMPILER==FLANG", in CMAKE builds the issue is probably masked by not setting an optimization level (unless -DCMAKE_BUILD_TYPE=Release is specified). Will test this later today.

**Comment by Diazonium** [comment]
BTW, is this issue related to the old LLVM Fortran frontend, or the new one (F18) that actually got merged into LLVM trunk recently?

**Comment by martin-frbg** [comment]
Old frontend i.e. original flang project. If the f18 frontend goes by the name "flang" as well we will probably need to differentiate by version number. (I do not recall seeing it mentioned in the LLVM10 release announcement, and have not tried building the very latest from their git)
EDIT: just confirmed my mid-february checkout of llvm10 still came without flang)

**Comment by leezu** [comment]
@Diazonium afaik the new f18 can't generate binaries yet, but only parse Fortran code. Not sure if it'd be ready by llvm 11 or later.

**Comment by leezu** [comment]
@martin-frbg is it possible that similar to https://github.com/flang-compiler/flang/issues/630 flang contains "correct" optimizations and OpenBLAS fortran code is buggy (causing the test failures)? Or do you think this is a bug in flang? If so, I'll open an issue in flang about the failing tests with O2 / O3.

**Comment by martin-frbg** [comment]
The fortran code in question is a direct copy of Reference-LAPACK (a.k.a "netlib LAPACK") and compiles fine with other compilers, so I'd think at best flang would have to be extra strict in its interpretation of the standard.

**Comment by leezu** [comment]
Thank you for clarifying. I opened https://github.com/flang-compiler/flang/issues/905

**Comment by martin-frbg** [comment]
Possibly a duplicate of their issue 599 (where it was suggested to add `-Kieee` to the compiler options) 
EDIT: that option appears to have been introduced only after their 201903029 binary release, but the AMD AOCC probably has it. I _think_ I tried it unsuccessfully before I came up with the -O1 workaround

**Comment by martin-frbg** [comment]
Interestingly these BLAS tests pass in a pure Reference-LAPACK build, so it is probably something related to argument passing between fortran and C functions (reference being all fortran)

**Comment by martin-frbg** [comment]
Actually this is now looking like some weird precision problem or a glitch in the comparison function, at least all of the tests that are failed for being "less than half accurate" show identical numbers in the "expected" and "computed" results printed. (Granted they are printed to limited precision, but I'd expect at least some of them to be way off the mark if the failure was real)

**Comment by martin-frbg** [comment]
Possibly just the calculation of the machine precision going wrong - OpenBLAS still has BLAS tests from version ~ 3.5.0 of the reference BLAS with an ad-hoc calculation of epsilon while current reference implementation uses the f90 epsilon intrinsic

**Comment by martin-frbg** [comment]
tests fixed by using f90 epsilon, but not the end of the story unfortunately as ctests still fail at -O2 even with the calculation of epsilon updated where necessary.

**Comment by martin-frbg** [comment]
Worse yet, running the LAPACK testsuite (make lapack-test) as opposed to just the BLAS and CBLAS tests goes completely haywire on Ryzen3900 at least with the AOCC flavor of clang/flang - lots of test failures and segmentation faults, ultimately culminating in an infinite loop in the single-precision complex tests. A mixed gcc+flang build fails in the same way while clang+gfortran works. Still need to rerun the test with flang20190329

**Comment by leezu** [comment]
@martin-frbg are these lapack-test issues with -O1 or -O2?

**Comment by martin-frbg** [comment]
O1, but I expect they will also be present at -O2.

**Comment by martin-frbg** [comment]
Dropping flang optimization level to O0 rather than just O1 works around this (and the ctest failures as well). Not sure if this can be narrowed down to a specific subset of optimizations ~~as my current understanding is that clang/flang has less fine-grained options than gcc/gfortran.~~
EDIT: found some listed here though probably not up to date https://stackoverflow.com/questions/15548023/clang-optimization-levels

**Comment by martin-frbg** [comment]
Meh. Half of the options on that list appear to be invalid with or without `-llvm` prepended and there does not appear to be an actual document of which is currently switched on at what level on the llvm website.

**Comment by Diazonium** [comment]
IMO, it is OK to have a "Known issues:" section in the release notes. If this is not readily fixable, or looks like a compiler bug, then perhaps it is not worth chasing the issue before 0.3.10.
Unless the bug was introduced recently, this went unreported for a very long time, suggesting that very few people actually use flang/AOCC-flang to compile OpenBLAS.

If a fix is found before 0.3.11, great, but in the long term flang will be eventually (LLVM 11? 12?) replaced with f18, which should be much better supported.

IMO 0.3.10 is already packed with many fixes and improvements, at some point a line should be drawn, and unsolved issues should go into a "Known issues" section of the release notes (as long as they are not regressions), instead of holding up the release

**Comment by martin-frbg** [comment]
lapack-test is back to normal at -O1 after correcting the -frecursive to -Mrecursive and adding -Kieee
At  -O2 however a significant number (2700 each) of the LAPACK complex and double-complex tests fail  with a spurious `parameter #4 had an illegal value` error on entry to CLASCL/ZLASCL (meaning CFROM was either zero or NaN - it is NaN when this happens). Trying -Mdaz to flush denormals to zero leads to a compiler error in the clang frontend.

**Comment by martin-frbg** [comment]
Latest in this saga is that with `-Mrecursive and -Kieee` added it is only AOCC going wrong after all, and adding `-fno-unroll-loops` to its options makes it pass the lapack tests at -O2 as well.
